cookbooks/bcpc/recipes/ssh.rb

```diff
@@ -74,118 +60,82 @@
-          Chef::Log.info("Saving #{key_type} for #{node[:fqdn]}")
-          vault_item[key_type] = ::File.read(private_key_path)
-          vault_item.save
-        end
-      end
-    else
-      Chef::Log.warn("No vault item found for ssh_host_keys/#{node[:fqdn]}!")
-    end
-  end
-  not_if { node[:fqdn] == get_bootstrap }
+def create_ssh_keys
+  vault_item = ChefVault::Item.new('ssh_host_keys', node[:fqdn])
+  vault_item.admins([get_bootstrap, node[:fqdn], 'admin'].join(','))
+  vault_item.search("fqdn:#{node[:fqdn]}")
+  vault_item.save
+rescue Exception => eee
+  Chef::Log.warn("Failed to create new vault ssh_host_keys/#{node[:fqdn]}")
+  Chef::Log.warn("Is this node an admin?\n#{eee}")
+  return nil
 end
 
-# If we successfully retrieved host keys, write files to disk and reload SSH.
-ruby_block 'write-ssh-secrets' do
-  block do
-    vault_item = node.run_state[:bcpc_ssh_host_keys]
-
-    if vault_item.is_a?(ChefVault::Item)
-      ssh_key_types.each do |key_type|
-        next unless vault_item[key_type]
-
-        private_key_path = "/etc/ssh/ssh_host_#{key_type}_key"
-        public_key_path = "/etc/ssh/ssh_host_#{key_type}_key.pub"
-
-        # This will contain ascii key data, or nil.
-        existing_key =
-          ::File.exist?(private_key_path) &&
-          ::File.read(private_key_path)
-
-        #
-        # If the private key saved on the server doesn't match the one
-        # on disk, replace it and regenerate the public key.
-        #
-        if vault_item[key_type] != existing_key
-          Chef::Log.info("Replacing #{key_type} host key for #{node[:fqdn]} " \
-                         "with version from vault (ssh_keys/#{node[:fqdn]})")
-
-          Chef::Resource::File.new(private_key_path,
-                                   node.run_context).tap do |ff|
-            ff.user 'root'
-            ff.group 'root'
-            ff.mode '0400'
-            ff.content vault_item[key_type]
-            ff.sensitive true if ff.respond_to?(:sensitive)
-            ff.run_action(:create)
-          end
-
-          Chef::Resource::Execute.new("generate-#{key_type}-public-key",
-                                      node.run_context).tap do |ee|
-            ee.umask 0222
-            ee.command "ssh-keygen -y -f #{private_key_path} > #{public_key_path}"
-            ee.run_action(:run)
-          end
-
-          #
-          # Since we have no run collection for our dynamically
-          # generated resources, we have to use the ruby_block itself
-          # to notify the SSH service.
-          #
-          self.notifies :restart, 'service[ssh]', :immediately
-          self.resolve_notification_references
-        end
+if node[:fqdn] != get_bootstrap
+  # First, try to retrieve existing host keys from the vault.
+  # If we fail to load from the vault, log the failure and create an item.
+  # If we also fail to create a vault item, this node may not be an admin.
+  vault_item = load_ssh_keys || create_ssh_keys
```

Comment:
This logic seems unobvious; what return do you expect from a successful `create_ssh_keys` call?